### PR TITLE
feature: localisation

### LINKED ISSUE #42: Localisation
Since we're building a government website, we need it to exist both in English and Hindi.

### PR DESCRIPTION
**Pull request by MohitMahipal**
Fixes #42

### CODE REVIEW

**Comment by GetPsyched** [comment]
Use this in the components and the 404 page instead of misc text

File: app/[lang]/dictionaries.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import 'server-only'
+
```
Comment (GetPsyched) [inline app/[lang]/dictionaries.ts:2]: Not required
```suggestion
```

File: app/[lang]/dictionaries.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+interface DictionaryInterface {
+  [k: string]: any;
+}
+
+const dictionaries: {[k: string]: () => Promise<DictionaryInterface> } = {
+  "en-GB": () => import('./dictionaries/en.json').then((module) => module.default),
+  "hi-IN": () => import('./dictionaries/hi.json').then((module) => module.default),
```
Comment (GetPsyched) [inline app/[lang]/dictionaries.ts:9]:
```suggestion
  "en-GB": () => import('./locales/en.json').then((module) => module.default),
  "hi-IN": () => import('./locales/hi.json').then((module) => module.default),
```

File: app/[lang]/dictionaries.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+}
+ 
+export const getDictionary = async (locale: string): Promise<DictionaryInterface> => {
```
Comment (GetPsyched) [inline app/[lang]/dictionaries.ts:12]:
```suggestion
export const getLocale = async (locale: string): Promise<DictionaryInterface> => {
```

**Review by GetPsyched — CHANGES_REQUESTED** [review]

File: app/[lang]/dictionaries/hi.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+    "products": {
+      "cart": "मोहित"
+    }
+  }
```
Comment (Maan2003) [inline app/[lang]/dictionaries/hi.json:5]: missing new line, ditto for some other files

File: middleware.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { headers as nextHeaders } from "next/headers";
+import { match } from "@formatjs/intl-localematcher";
+import Negotiator from "negotiator";
+
+let locales: string[] = ["en-GB", "hi-IN"];
```
Comment (Maan2003) [inline middleware.ts:5]: const

File: middleware.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+function getLocale(request: any): string {
+    let languages = new Negotiator({
+        headers: {
+      "accept-language": nextHeaders().get("Accept-Language") as any
+    }}).languages();
```
Comment (Maan2003) [inline middleware.ts:11]: format

File: app/[lang]/dictionaries.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import 'server-only'
+
+interface DictionaryInterface {
```
Comment (Maan2003) [inline app/[lang]/dictionaries.ts:3]:
```suggestion
interface Dictionary {
```

File: app/[lang]/dictionaries.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  [k: string]: any;
```
Comment (Maan2003) [inline app/[lang]/dictionaries.ts:4]:
```suggestion
  [k: string]: string;
```

File: app/[lang]/dictionaries.ts [hunk 9]
```diff
@@ -1,15 +1,21 @@
-import 'server-only'
+import 'server-only';
+
+type Obj<T> = {
+  [k: string]: T;
+};
 
 interface DictionaryInterface {
-  [k: string]: any;
+  [k: string]: () => Promise<Obj<Obj<string>>>;
```
Comment (Maan2003) [inline app/[lang]/dictionaries.ts:8]: what, looks incorrect

File: app/[lang]/dictionaries/en.json [hunk 10]
```diff
@@ -1,5 +1,5 @@
 {
-    "products": {
-      "cart": "Mohit"
-    }
-  }
\ No newline at end of file
+  "products": {
+    "cart": "Mohit"
+  }
```
Comment (Maan2003) [inline app/[lang]/dictionaries/en.json:4]: why not, you get a flat type then
```suggestion
  "products.cart": "Mohit",
```

File: app/[lang]/dictionaries.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import 'server-only';
+
+type Obj<T> = {
+  [k: string]: T;
+};
+
+interface DictionaryInterface {
+  [k: string]: () => Promise<Obj<Obj<string>>>;
```
Comment (Maan2003) [inline app/[lang]/dictionaries.ts:8]: better to use inferred types here. so you will get completions on the items in dictionary

File: components/work-in-progress.tsx [hunk 12]
```diff
@@ -1,8 +1,11 @@
-export default function WorkInProgress() {
+import { getLocale } from './dictionaries';
+
+export default async function WorkInProgress({ lang }: { lang: string }) {
+  const dict = await getLocale(lang);
   return (
-    <div className="m-auto max-w-fit text-center">
```
Comment (GetPsyched) [inline components/work-in-progress.tsx:3]: Why are we removing the styling??

**Review by Maan2003 — APPROVED** [review]

File: components/work-in-progress.tsx (hunk 12 above)
Comment (MohitMahipal) [reply components/work-in-progress.tsx:3]: fixed

File: middleware.ts (hunk 6 above)
Comment (Maan2003) [inline middleware.ts:11]: why is `npm run format` not fixing this

cc @GetPsyched

**Comment by Maan2003** [comment]
feels like prop drilling

**Comment by GetPsyched** [comment]
> feels like prop drilling

Not passing those props. NextJS directly injects props to page.tsx files which exist in the route path

**Comment by Maan2003** [comment]
what, i imagine every component needs translation, so you have to pass this lang prop to every component

--

this is not how i have seen l10n done,  usually there is function `t` that returns a translate version. `t` doesn't take lang as a paramater, it is somehow global state.

**Comment by GetPsyched** [comment]
> `t` doesn't take lang as a paramater, it is somehow global state.

That somehow is with the use of the `useContext` hook. Which is a feature that requires the components to be client-side. Beats the purpose of SSR. If there's an alternative, I'm open to it.

**Review by GetPsyched — APPROVED** [review]
A bunch of things remain untranslated but that's out of scope of this PR.

**Pull request merged**